networkx/algorithms/polynomials/tutte.py

```diff
@@ -0,0 +1,180 @@
+    the graph obtained from `G` by deleting edge `e`, `G/e` the graph obtained
+    from `G` by contracting edge `e`, `k(G)` the number of cut-edges of `G`,
+    and `l(G)` the number of loops of `G`:
+
+    .. math::
+        T_G(x, y) = \begin{cases}
+    	   x^{k(G)} y^{l(G)}, & \text{if all edges are cut-edges or loops} \\
+           T_{G-e}(x, y) + T_{G/e}(x, y), & \text{otherwise, for $e$ not a cut-edge or loop}
+        \end{cases}
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    simplify : bool
+        determines if the returned expression will be in simplified form
+
+    Returns
+    -------
+    instance of :class:`sympy.core.add.Add`
+        A Sympy expression representing the Tutte polynomial for `G`.
+
+    Examples
+    --------
+    >>> C = nx.cycle_graph(5)
+    >>> nx.tutte_polynomial(C)
+    x**4 + x**3 + x**2 + x + y
+
+    >>> D = nx.diamond_graph()
+    >>> nx.tutte_polynomial(D)
+    x**3 + 2*x**2 + 2*x*y + x + y**2 + y
+
+    Notes
+    -----
+    This function implements the deletion-contraction recurrence (Definition 3,
+    described above). Edge contraction is defined and deletion-contraction is
+    introduced in [3]_. Combinatorial meaning of the coefficients is introduced
+    in [4]_. Universality, properties, and applications are discussed in [5]_.
+
+    See Also
+    --------
+    _get_cut_edges
+
+    References
+    ----------
+    .. [1] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 14
+    .. [2] Y. Shi, M. Dehmer, X. Li, I. Gutman,
+       "Graph Polynomials," p. 46
+    .. [3] D. B. West,
+       "Introduction to Graph Theory," p. 84
+    .. [4] G. Coutinho,
+       "A brief introduction to the Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://homepages.dcc.ufmg.br/~gabriel/seminars/coutinho_tuttepolynomial_seminar.pdf
+    .. [5] J. A. Ellis-Monaghan, C. Merino,
+       "Graph polynomials and their applications I: The Tutte polynomial"
+       Structural Analysis of Complex Networks, 2011
+       https://arxiv.org/pdf/0803.3079.pdf
+    """
+    import sympy
+
+    G = nx.MultiGraph(G)
+
+    components = list(nx.connected_components(G))
+    if len(components) > 1:
+        t = 1
+        for component in components:
+            t *= tutte_polynomial(G.subgraph(component), False)
+        return sympy.simplify(t)
+
+    cut_edges = _get_cut_edges(G)
+    loops = [e for e in G.edges if e[0] == e[1]]
+    edges_not_cuts_loops = [i for i in G.edges if i not in cut_edges and i not in loops]
+
+    if not edges_not_cuts_loops:
+        return sympy.Symbol("x") ** len(cut_edges) * sympy.Symbol("y") ** len(loops)
+
+    e = edges_not_cuts_loops[0]
+    # deletion-contraction
+    C = nx.contracted_edge(G, e, self_loops=True)
+    contraction_loops = [i for i in C.edges if i[0] == i[1] and i not in G.edges]
+    if contraction_loops:
+        C.remove_edge(*contraction_loops[0])
```

Comment:
Shortcut this loop:           
```suggestion
    for e in C.edges if e[0] == e[1] and e not in G.edges:
        C.remove_edge(*e)
```
But actually --- why is this needed. Can you just use `self_loops=False` in the `contracted_edge` function?

Reply:
@dschult Good point, I will shortcut the loop as you describe.

This is needed because using `self_loops=False` in the `contracted_edge` function prohibits all loops arising from edge contraction. In this computation, we want to permit loops, but not arising from the particular edge we are contracting. The disagreement between these two approaches arises when you contract an edge with multiplicity greater than one.

For instance, suppose we have a three-vertex graph where one edge has multiplicity two:

```python
G = nx.MultiGraph()
G.add_edge(1, 2)
G.add_edge(1, 2)
G.add_edge(2, 3)
G.add_edge(1, 3)
```

Contracting one of the edges (1, 2) using `self_loops=False` in the `contracted_edge` function produces a graph with two vertices, two edges, and no loops. What we are looking for, however, is a graph with two vertices and three edges (one of which is a loop).